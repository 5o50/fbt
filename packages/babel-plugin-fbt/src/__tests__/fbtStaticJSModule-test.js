/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 *
 * @emails oncall+i18n_fbt_js
 * @format
 */

jest.autoMockOff();

const {
  jsCodeFbtCallSerializer,
  payload,
  snapshotTransform,
  transform,
  withFbtRequireStatement,
} = require('../FbtTestUtil');
const {TestUtil} = require('fb-babel-plugin-utils');

function runTest(data, extra) {
  var expected = data.output;
  var actual = transform(data.input, extra);
  TestUtil.assertSourceAstEqual(expected, actual);
}

expect.addSnapshotSerializer(jsCodeFbtCallSerializer);

describe('fbt preserveWhitespace argument', () => {
  // TODO: T38897324 (#32) Fix space normalization.
  // Here we are intentionally testing for the wrong behavior. We will come
  // back and update the expected output after we fix space normalization.
  describe('should NOT preserve whitespaces that do not neighbor raw text', () => {
    const snapshotTestData = {
      'jsx elements and raw text': {
        input: withFbtRequireStatement(`
          var x =
            <fbt desc="d">
              <span>
                Where do
              </span>
              <b>spaces</b>
              <i>go?</i>
              Good
              <i>question</i>
              !
            </fbt>;
        `),
      },
      'jsx elements with string variation arguments': {
        input: withFbtRequireStatement(`
          var x =
            <fbt desc="d">
              <span>
                There should be
              </span>
              <b>
                <fbt:plural
                  many="spaces"
                  showCount="ifMany"
                  count={this.state.ex1Count}>
                  a space
                </fbt:plural>
              </b>
              !
            </fbt>;
        `),
      },
    };
    TestUtil.testSection(snapshotTestData, snapshotTransform, {
      matchSnapshot: true,
    });
  });

  it('should preserve whitespace in text when requested', () => {
    runTest({
      input: withFbtRequireStatement(
        'var x = fbt("two\\nlines", "one line", {preserveWhitespace:true});',
      ),
      output: withFbtRequireStatement(
        `var x = fbt._(${payload({
          jsfbt: {
            t: {
              desc: 'one line',
              text: 'two\nlines',
            },
            m: [],
          },
        })})`,
      ),
    });

    runTest({
      input: withFbtRequireStatement(
        'var x = fbt("two  spaces", "one space", {preserveWhitespace:true});',
      ),
      output: withFbtRequireStatement(
        `var x = fbt._(${payload({
          jsfbt: {
            t: {
              desc: 'one space',
              text: 'two  spaces',
            },
            m: [],
          },
        })})`,
      ),
    });
  });

  it('should preserve whitespace in desc when requested', () => {
    runTest({
      input: withFbtRequireStatement(
        `var x = fbt('one line', 'two\\nlines', {preserveWhitespace: true});`,
      ),

      output: withFbtRequireStatement(
        `var x = fbt._(
            ${payload({
              jsfbt: {
                t: {
                  desc: 'two\nlines',
                  text: 'one line',
                },
                m: [],
              },
            })},
          );`,
      ),
    });

    runTest({
      input: withFbtRequireStatement(
        `var x = fbt('one space', 'two  spaces', {preserveWhitespace: true});`,
      ),
      output: withFbtRequireStatement(
        `var x = fbt._(
            ${payload({
              jsfbt: {
                t: {
                  desc: 'two  spaces',
                  text: 'one space',
                },
                m: [],
              },
            })},
          );`,
      ),
    });
  });

  it('should coalesce whitespace in text when not requested', () => {
    runTest({
      input: withFbtRequireStatement(
        `var x = fbt('two  spaces', 'one space', {preserveWhitespace: false});`,
      ),
      output: withFbtRequireStatement(
        `var x = fbt._(
            ${payload({
              jsfbt: {
                t: {
                  desc: 'one space',
                  text: 'two spaces',
                },
                m: [],
              },
            })},
          );`,
      ),
    });

    runTest({
      input: withFbtRequireStatement(
        `var x = fbt('two\\nlines', 'one line', {preserveWhitespace: false});`,
      ),
      output: withFbtRequireStatement(
        `var x = fbt._(
            ${payload({
              jsfbt: {
                t: {
                  desc: 'one line',
                  text: 'two lines',
                },
                m: [],
              },
            })},
          );`,
      ),
    });
  });

  it('should coalesce whitespace in desc when not requested', () => {
    runTest({
      input: withFbtRequireStatement(
        `var x = fbt('one line', 'two\\nlines', {preserveWhitespace: false});`,
      ),
      output: withFbtRequireStatement(
        `var x = fbt._(
            ${payload({
              jsfbt: {
                t: {
                  desc: 'two lines',
                  text: 'one line',
                },
                m: [],
              },
            })},
          );`,
      ),
    });

    runTest({
      input: withFbtRequireStatement(
        `var x = fbt('one space', 'two spaces', {preserveWhitespace: false});`,
      ),
      output: withFbtRequireStatement(
        `var x = fbt._(
            ${payload({
              jsfbt: {
                t: {
                  desc: 'two spaces',
                  text: 'one space',
                },
                m: [],
              },
            })},
          );`,
      ),
    });
  });
});
