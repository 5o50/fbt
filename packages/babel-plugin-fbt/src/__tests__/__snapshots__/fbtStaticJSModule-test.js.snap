// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fbt preserveWhitespace argument should NOT preserve whitespaces that do not neighbor raw text jsx elements and raw text 1`] = `
const fbt = require("fbt");

var x = fbt._(
  /* __FBT__ start */ {
    jsfbt: {
      t: {
        desc: "d",
        text: "{=Where do}{=spaces}{=go?} Good {=question} !",
        tokenAliases: {
          "=Where do": "=m0",
          "=spaces": "=m1",
          "=go?": "=m2",
          "=question": "=m4",
        },
      },
      m: [],
    },
    project: "",
  } /* __FBT__ end */,
  [
    fbt._implicitParam(
      "=m0",
      /*#__PURE__*/ React.createElement(
        "span",
        null,
        fbt._(
          /* __FBT__ start */ {
            jsfbt: {
              t: {
                desc:
                  'In the phrase: "{=Where do}{=spaces}{=go?} Good {=question} !"',
                text: "Where do",
              },
              m: [],
            },
            project: "",
          } /* __FBT__ end */
        )
      )
    ),
    fbt._implicitParam(
      "=m1",
      /*#__PURE__*/ React.createElement(
        "b",
        null,
        fbt._(
          /* __FBT__ start */ {
            jsfbt: {
              t: {
                desc:
                  'In the phrase: "{=Where do}{=spaces}{=go?} Good {=question} !"',
                text: "spaces",
              },
              m: [],
            },
            project: "",
          } /* __FBT__ end */
        )
      )
    ),
    fbt._implicitParam(
      "=m2",
      /*#__PURE__*/ React.createElement(
        "i",
        null,
        fbt._(
          /* __FBT__ start */ {
            jsfbt: {
              t: {
                desc:
                  'In the phrase: "{=Where do}{=spaces}{=go?} Good {=question} !"',
                text: "go?",
              },
              m: [],
            },
            project: "",
          } /* __FBT__ end */
        )
      )
    ),
    fbt._implicitParam(
      "=m4",
      /*#__PURE__*/ React.createElement(
        "i",
        null,
        fbt._(
          /* __FBT__ start */ {
            jsfbt: {
              t: {
                desc:
                  'In the phrase: "{=Where do}{=spaces}{=go?} Good {=question} !"',
                text: "question",
              },
              m: [],
            },
            project: "",
          } /* __FBT__ end */
        )
      )
    ),
  ]
);

`;

exports[`fbt preserveWhitespace argument should NOT preserve whitespaces that do not neighbor raw text jsx elements with string variation arguments 1`] = `
var fbt_sv_arg_0;

const fbt = require("fbt");

var x =
  ((fbt_sv_arg_0 = fbt._plural(this.state.ex1Count, "number")),
  fbt._(
    /* __FBT__ start */ {
      jsfbt: {
        t: {
          "*": {
            desc: "d",
            text: "{=There should be}{=[number] spaces} !",
            tokenAliases: {
              "=There should be": "=m0",
              "=[number] spaces": "=m1",
            },
          },
          _1: {
            desc: "d",
            text: "{=There should be}{=a space} !",
            tokenAliases: { "=There should be": "=m0", "=a space": "=m1" },
          },
        },
        m: [{ token: "number", type: 2, singular: true }],
      },
      project: "",
    } /* __FBT__ end */,
    [
      fbt_sv_arg_0,
      fbt._implicitParam(
        "=m0",
        /*#__PURE__*/ React.createElement(
          "span",
          null,
          fbt._(
            /* __FBT__ start */ {
              jsfbt: {
                t: {
                  "*": {
                    desc:
                      'In the phrase: "{=There should be}{=[number] spaces} !"',
                    text: "There should be",
                  },
                  _1: {
                    desc: 'In the phrase: "{=There should be}{=a space} !"',
                    text: "There should be",
                  },
                },
                m: [{ token: "number", type: 2, singular: true }],
              },
              project: "",
            } /* __FBT__ end */,
            [fbt_sv_arg_0]
          )
        )
      ),
      fbt._implicitParam(
        "=m1",
        /*#__PURE__*/ React.createElement(
          "b",
          null,
          fbt._(
            /* __FBT__ start */ {
              jsfbt: {
                t: {
                  "*": {
                    desc:
                      'In the phrase: "{=There should be}{=[number] spaces} !"',
                    text: "{number} spaces",
                  },
                  _1: {
                    desc: 'In the phrase: "{=There should be}{=a space} !"',
                    text: "a space",
                  },
                },
                m: [{ token: "number", type: 2, singular: true }],
              },
              project: "",
            } /* __FBT__ end */,
            [fbt_sv_arg_0]
          )
        )
      ),
    ]
  ));

`;
